### PR TITLE
42 key facts

### LINKED ISSUE #42: Key facts
Adjust existing block with new typography, which will fix #31. We’re adding an image to the block as well. In this case, the alignment of the key facts will be different. 

The content makes use of available space as much as possible
- On mobile and tablet, the facts are [placed below the visual](https://www.figma.com/file/yv0VtHTvMoGPqkv0RAYcVp/PDP?type=design&node-id=1091-22224&mode=design&t=KmUPP5UMoxagXFmW-4)
- When possible, the facts are [placed in one row](https://www.figma.com/file/yv0VtHTvMoGPqkv0RAYcVp/PDP?type=design&node-id=1091-23179&mode=design&t=KmUPP5UMoxagXFmW-4).
- On desktop, the facts are shown [next to the visual](https://www.figma.com/file/yv0VtHTvMoGPqkv0RAYcVp/PDP?type=design&node-id=3781-47910&mode=design&t=KmUPP5UMoxagXFmW-4).

**Technical notes**
- Reuse the same block (Key facts), refactor the component and adjust the typography
- If an image is added, readjust the content to fit the image; ELSE, content should be arranged is today

### PR DESCRIPTION
**Pull request by TomaszDziezykNetcentric**
Please always provide the [GitHub issue(s)](../issues) your PR is for, as well as test URLs where your change can be observed (before and after):

Fix #42 

Test URLs:
- Before: https://main--vg-volvotrucks-us-rd--netcentric.hlx.page/drafts/tdziezyk/v2/key-facts
- After: https://42-key-facts--vg-volvotrucks-us-rd--netcentric.hlx.page/drafts/tdziezyk/v2/key-facts

### CODE REVIEW

File: blocks/key-facts/key-facts.js [hunk 1]
```diff
@@ -6,40 +9,95 @@ function stripEmptyTags(main, child) {
...
+      `;
+    }
+
+    const numberFragment = document.createRange().createContextualFragment(newValue);
+    value.innerHTML = '';
+    value.append(...numberFragment.childNodes);
+    value.classList.add(`${blockName}__number-wrapper`, 'h2');
+
+    if (!v2 && text) {
+      const unit = createElement('strong', [`${blockName}__unit`, 'sub-2']);
```
Comment (Lakshmishri) [inline blocks/key-facts/key-facts.js:60]: `createElement ` is now changed to accept classes as options, can you change this across file

File: styles/styles.css [hunk 2]
```diff
@@ -1199,6 +1199,54 @@ main.blue-contract .section.section-with-title p {
   width: 1em;
 }
 
+/* Headings classes styles */
+.redesign-v2 .h2 {
+  font-family: var(--ff-volvo-novum);
+  font-size: 56px;
```
Comment (Lakshmishri) [inline styles/styles.css:1205]: can use `var(--f-heading-2-font-size)`

File: styles/styles.css [hunk 3]
```diff
@@ -1199,6 +1199,54 @@ main.blue-contract .section.section-with-title p {
...
+  font-style: normal;
+  font-weight: 500;
+  line-height: 125%;
+  letter-spacing: 0.25px;
+  font-feature-settings: 'clig' off, 'liga' off;
+}
+
+.redesign-v2 .h5 {
+  font-family: var(--ff-volvo-novum);
+  font-size: 24px;
```
Comment (Lakshmishri) [inline styles/styles.css:1225]: Use `var(--f-heading-5-font-size)`

File: styles/styles.css [hunk 4]
```diff
@@ -1199,6 +1199,54 @@ main.blue-contract .section.section-with-title p {
...
+  font-style: normal;
+  line-height: 125%;
+  font-feature-settings: 'clig' off, 'liga' off;
+}
+
+.redesign-v2 .h6 {
+  font-family: var(--ff-volvo-novum-medium);
+  font-size: 20px;
```
Comment (Lakshmishri) [inline styles/styles.css:1233]: use` var(--f-heading-6-font-size)`

File: styles/styles.css [hunk 5]
```diff
@@ -1199,6 +1199,54 @@ main.blue-contract .section.section-with-title p {
...
+  font-style: normal;
+  line-height: 125%;
+  letter-spacing: 0.15px;
+  font-feature-settings: 'clig' off, 'liga' off;
+
+}
+
+.redesign-v2 .sub-2 {
+  font-family: var(--ff-volvo-novum-medium);
+  font-size: 14px;
```
Comment (Lakshmishri) [inline styles/styles.css:1243]: use `var(--f-subtitle-2-font-size)`

**Review by Lakshmishri — CHANGES_REQUESTED** [review]
Should the video also have border radius @cogniSyb , can you confirm ?

The fonts for h2 look generic , they` font-size: 56px` which isn't as per the design for [smaller screen size](https://www.figma.com/file/yv0VtHTvMoGPqkv0RAYcVp/PDP?type=design&node-id=1091-23190&mode=dev) . Is it expected ?

**Review by kesiah — CHANGES_REQUESTED** [review]

File: blocks/key-facts/key-facts.js (hunk 1 above)
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.js]: Good catch :) 
fixed

File: styles/styles.css (hunk 2 above)
Comment (TomaszDziezykNetcentric) [reply styles/styles.css]: Fixed

File: styles/styles.css (hunk 3 above)
Comment (TomaszDziezykNetcentric) [reply styles/styles.css]: done

File: styles/styles.css (hunk 5 above)
Comment (TomaszDziezykNetcentric) [reply styles/styles.css:1247]: done

File: styles/styles.css [hunk 6]
```diff
@@ -1201,6 +1201,56 @@ main.blue-contract .section.section-with-title p {
...
+  /* font-size: 56px; */
+  font-size: var(--f-heading-2-font-size);
+  font-style: normal;
+  font-weight: 300;
+  line-height: 110%;
+  letter-spacing: -0.5px;
+  font-feature-settings: 'clig' off, 'liga' off;
+}
+
+.redesign-v2 h4 {
```
Comment (kesiah) [inline styles/styles.css:1217]: Why the rest are `.redesign-v2 .h5` and this one is changing the tag directly?

File: styles/styles.css [hunk 7]
```diff
@@ -1199,6 +1199,54 @@ main.blue-contract .section.section-with-title p {
   width: 1em;
 }
 
+/* Headings classes styles */
+.redesign-v2 .h2 {
```
[the remark this answers is not in the document]
Comment (kesiah) [reply styles/styles.css:1205]: Yes, I agree on doing that, but you should also use the variable for line-height and letter spacing.

File: styles/styles.css [hunk 8]
```diff
@@ -1201,6 +1201,56 @@ main.blue-contract .section.section-with-title p {
...
+
+  /* font-size: 56px; */
+  font-size: var(--f-heading-2-font-size);
+  font-style: normal;
+  font-weight: 300;
+  line-height: 110%;
+  letter-spacing: -0.5px;
+  font-feature-settings: 'clig' off, 'liga' off;
+}
+
```
Comment (kesiah) [inline styles/styles.css:1216]: Missing `.redesign-v2 .h3` code

File: styles/styles.css [hunk 9]
```diff
@@ -1201,6 +1201,56 @@ main.blue-contract .section.section-with-title p {
...
+  font-family: var(--ff-volvo-novum-medium);
+  font-size: var(--f-heading-6-font-size);
+  font-style: normal;
+  line-height: 125%;
+  letter-spacing: 0.15px;
+  font-feature-settings: 'clig' off, 'liga' off;
+
+}
+
+.redesign-v2 .sub-2 {
```
Comment (kesiah) [inline styles/styles.css:1245]: What's sub-2? Can we rename it to `subtitle-2`? If we want to have generic class names, they need to be clear.
And if it's a generic style, why we are not using the CSS variables for the font size / letter spacing / line-hegiht? in this case `--f-subtitle-2-font-size`

**Review by kesiah — CHANGES_REQUESTED** [review]

File: styles/styles.css (hunk 7 above)
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply styles/styles.css:1213]: Updated to css variables

File: styles/styles.css (hunk 6 above)
Comment (TomaszDziezykNetcentric) [reply styles/styles.css]: It is a typo :)
Fixed

File: styles/styles.css (hunk 8 above)
Comment (TomaszDziezykNetcentric) [reply styles/styles.css:1228]: Added

File: styles/styles.css (hunk 9 above)
Comment (TomaszDziezykNetcentric) [reply styles/styles.css]: Not problem, changed to `subtitle-2`

File: styles/styles.css [hunk 10]
```diff
@@ -1201,6 +1201,71 @@ main.blue-contract .section.section-with-title p {
...
+  font-feature-settings: 'clig' off, 'liga' off;
+}
+
+.redesign-v2 .h6 {
+  font-family: var(--ff-volvo-novum-medium);
+  font-size: var(--f-heading-6-font-size);
+  line-height: var( --f-heading-6-line-height);
+  letter-spacing: var(--f-heading-6-letter-spacing);
+  font-feature-settings: 'clig' off, 'liga' off;
+
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply styles/styles.css]: The empty line was removed.

**Comment by TomaszDziezykNetcentric** [comment]
> image Is the line after Range normal? I cannot see it in the designs, there are so many variants that I cannot confirm :(

The line is a variant of the old design. I don't know if this is needed anymore...
It is added only when the variant `trailing-line` is added. We didn't create the v2 version of `key-facts` so I made all the changes to be backward compatible.

File: blocks/key-facts/key-facts.js [hunk 11]
```diff
@@ -6,40 +9,95 @@ function stripEmptyTags(main, child) {
...
+    stripEmptyTags(el, paragraph);
+  });
+
+  const preSubheading = el.querySelector(':scope > :not(strong, i):first-child');
+  preSubheading?.classList.add('subtitle-2');
+
+  // find and split number/unit
+  const value = el.querySelector('strong:only-child');
+  // separate number and unit. e.g. up | +31% | Freight efficiency, or 120,000 | psi,
+  const [, prenumber, number, postnumber, text] = value.innerHTML.match('([\\+,\\-]*)([0-9,\\.,]+)([\\%]*) *(.*)');
```
Comment (kesiah) [inline blocks/key-facts/key-facts.js:38]: Nice to have: testing with test cases :)

**Review by kesiah — APPROVED** [review]

File: blocks/key-facts/key-facts.js (hunk 11 above)
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.js:38]: Well in my opinion it will be nice to get rid of the regexp :)
But we don't know on which pages it is used.

**Review by markovukiceviccn — APPROVED** [review]

File: blocks/key-facts/key-facts.js [hunk 12]
```diff
@@ -6,40 +9,95 @@ function stripEmptyTags(main, child) {
...
+    }
+
+    const numberFragment = document.createRange().createContextualFragment(newValue);
+    value.innerHTML = '';
+    value.append(...numberFragment.childNodes);
+    value.classList.add(`${blockName}__number-wrapper`, 'h2');
+
+    if (!v2 && text) {
+      const unit = createElement('strong', { classes: [`${blockName}__unit`, 'ssubtitle-2'] });
+      unit.innerText = text;
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.js:65]: Please check now
obraz

File: blocks/key-facts/key-facts.js [hunk 13]
```diff
@@ -6,40 +9,95 @@ function stripEmptyTags(main, child) {
...
+  preSubheading?.classList.add('subtitle-2');
+
+  // find and split number/unit
+  const value = el.querySelector('strong:only-child');
+  // separate number and unit. e.g. up | +31% | Freight efficiency, or 120,000 | psi,
+  const [, prenumber, number, postnumber, text] = value.innerHTML.match('([\\+,\\-]*)([0-9,\\.,]+)([\\%]*) *(.*)');
+  let newValue = null;
+
+  if (value) {
+    if (v2) {
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.js]: I added limitation for 12 characters. If the text is longer the text is smaller. Please check the last item on the image.
obraz

File: blocks/key-facts/key-facts.css [hunk 14]
```diff
@@ -2,138 +2,127 @@
...
 
-/* icon */
-.key-facts-wrapper .key-fact i.fa {
-  font-size: 30px;
-  margin-bottom: 8px;
+.key-facts__number-wrapper {
+  display: flex;
+  align-items: center;
+  justify-content: center;
+  color: var(--c-cta-blue-default);
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.css:37]: Updated
obraz

File: blocks/key-facts/key-facts.js [hunk 15]
```diff
@@ -6,40 +9,95 @@ function stripEmptyTags(main, child) {
...
+      `;
+    }
+
+    const numberFragment = document.createRange().createContextualFragment(newValue);
+    value.innerHTML = '';
+    value.append(...numberFragment.childNodes);
+    value.classList.add(`${blockName}__number-wrapper`, 'h2');
+
+    if (!v2 && text) {
+      const unit = createElement('strong', { classes: [`${blockName}__unit`, 'ssubtitle-2'] });
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.js]: Fixed, thanks :)

File: blocks/key-facts/key-facts.js [hunk 16]
```diff
@@ -6,40 +9,137 @@ function stripEmptyTags(main, child) {
...
+      if (isImageCell) {
+        block.classList.add(`${blockName}--with-media`);
+        row.classList.add(`${blockName}__row--with-media`);
+        col.classList.add(`${blockName}__image-cell`);
+
+        return;
+      }
+
+      const v2 = block.classList.contains(`${blockName}--with-media`);
+      keyFactsColumns(col, v2);
```
Comment (kesiah) [inline blocks/key-facts/key-facts.js:142]: `v2` variable is a bit confusing, at least for me, maybe we could rename it to `isWithMedia`?

File: blocks/key-facts/key-facts.css [hunk 17]
```diff
@@ -2,138 +2,175 @@
...
-.key-facts-wrapper .key-fact i.fa {
-  font-size: 30px;
-  margin-bottom: 8px;
-}
-
-.key-facts-wrapper .key-fact .icon,
-.key-facts-wrapper .key-fact .icon svg {
-  height: 48px;
-  width: 48px;
+.key-facts .key-facts__main-text-wrapper {
```
Comment (kesiah) [inline blocks/key-facts/key-facts.css:33]: Could you add `text-wrap: balance;`? when the text is long, looks better ;)

**Review by kesiah — CHANGES_REQUESTED** [review]

File: blocks/key-facts/key-facts.css [hunk 18]
```diff
@@ -2,138 +2,175 @@
...
-    width: 72px;
+  .key-facts .key-facts__main-text-wrapper--small {
+    font-size: var(--f-heading-3-5-font-size);
+    letter-spacing: var(--f-heading-3-5-letter-spacing);
+    line-height: var(--f-heading-3-5-line-height);
+  }
+
+  .key-facts--with-media {
+    flex-direction: row;
+    gap: 40px;
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.css]: Fixed

File: blocks/key-facts/key-facts.css [hunk 19]
```diff
@@ -2,138 +2,175 @@
...
-    padding-bottom: 80px;
+    margin: 40px 0;
+    justify-content: space-between;
+    width: auto;
+    max-width: unset;
+  }
+
+  .key-facts--with-media .key-facts__row:not(.key-facts__row--with-media) {
+    min-width: 27%;
+    width: 27%;
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.css]: Changed

File: blocks/key-facts/key-facts.css [hunk 20]
```diff
@@ -2,138 +2,175 @@
...
+.key-facts__image-cell img {
+  border-radius: 8px;
+  width: 100%;
+  height: auto;
+  overflow: hidden;
+}
+
+.key-facts__row .block.embed,
+.key-facts__row .embed-video {
+  width: 100%;
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.css:103]: Done

File: blocks/key-facts/key-facts.css [hunk 21]
```diff
@@ -2,138 +2,175 @@
...
 
-@media screen and (min-width: 576px) {
-  .key-facts-wrapper .key-fact {
-    flex: 0 0 16.6667%;
+.key-facts__image-cell {
+  width: 100%;
+}
+
+.key-facts__image-cell img {
+  border-radius: 8px;
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.css]: Changed

File: blocks/key-facts/key-facts.css (hunk 17 above)
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.css:33]: Nice, done

File: blocks/key-facts/key-facts.css [hunk 22]
```diff
@@ -82,12 +83,16 @@
...
+  margin: 0 -16px;
+}
+
 .key-facts__image-cell {
   width: 100%;
 }
 
 .key-facts__image-cell img {
-  border-radius: 8px;
+  border-radius: var(--border-radius);
```
[the remark this answers is not in the document]
Comment (TomaszDziezykNetcentric) [reply blocks/key-facts/key-facts.css]: Done

BTW the images used in the document has rounded corners, it is not rounded by CSS anymore.

**Review by kesiah — APPROVED** [review]

**Review by Lakshmishri — APPROVED** [review]

**Pull request merged**